### PR TITLE
Resource auth

### LINKED ISSUE #58: Authentication endpoint for the INTERNET access

### PR DESCRIPTION
**Pull request by cytadela8**
This closes #58

### CODE REVIEW

File: wwwapp/admin.py [hunk 1]
```diff
@@ -2,6 +2,7 @@
 from django.contrib.auth.admin import UserAdmin
 from django.contrib.auth.models import User
 
+from wwwapp.models import ResourceYearPermission
```
[the remark this answers is not in the document]
Comment (cytadela8) [reply wwwapp/admin.py]: I will probably change everything. This line was added automatically by PyCharm.

File: wwwapp/models.py [hunk 2]
```diff
@@ -281,3 +289,23 @@ def result_in_percent(self):
 
     class Meta:
         unique_together = [('workshop', 'participant')]
+
+
+class ResourceYearPermission(models.Model):
+    display_name = models.CharField(max_length=32, blank=True)
```
[the remark this answers is not in the document]
Comment (cytadela8) [reply wwwapp/models.py]: It's displayed in the navbar. Long names will not look good there. (probably more then 20 characters is too much)

[the remark this answers is not in the document]
Comment (krzys-h) [reply wwwapp/models.py]: Articles are on the navbar as well and they have 50 - not that it matters too much, I just like things like that to be consistent everywhere

File: wwwapp/settings.py [hunk 3]
```diff
@@ -244,3 +244,7 @@
 # Max amount of points that a participant can get during qualification (relative to configured max_points for a given workshop)
 # This allows to give some people bonus points above 100%
 MAX_POINTS_PERCENT = 200
+
+# To protect against escaping INTERNET directory through symlinks and accessing
+# other files set this to INTERNET mount path on application server
+INTERNET_MNT_PATH = None
```
[the remark this answers is not in the document]
Comment (cytadela8) [reply wwwapp/settings.py]: This is to prevent espacing via symlinks from one resource folder to other. The files will be served by the same server as Django app so giving this information to Django won't be a problem. It's not feasible to use nginx for this separation as it's not aware of different resource directories. (only of the directory holding all data)

File: wwwapp/models.py [hunk 4]
```diff
@@ -22,6 +21,15 @@ class UserProfile(models.Model):
...
                or Workshop.objects.filter(type__year=year, lecturer=self, status='Z').exists()
+
+    def participation_years(self):
+        return self.participant_years().union(self.lecturer_years())
+
+    def participant_years(self):
+        return set([profile.year for profile in self.workshop_profile.filter(status=WorkshopUserProfile.STATUS_ACCEPTED)])
+
+    def lecturer_years(self):
+        return set([workshop.type.year for workshop in Workshop.objects.filter(lecturer=self, status='Z')])
```
[the remark this answers is not in the document]
Comment (cytadela8) [reply wwwapp/models.py]: We would need to introduce the other constant :P

File: wwwapp/views.py [hunk 5]
```diff
@@ -588,3 +595,31 @@ def page(request):
...
+@login_required()
+def resource_auth_view(request):
+    if request.user.has_perm('wwwapp.access_all_resources'):
+        return HttpResponse("Glory to WWW and the ELITARNY MIMUW!!!")
+
+    user_profile = UserProfile.objects.get(user=request.user)
+
+    path = request.META.get('HTTP_X_ORIGINAL_URI', '')
+    if settings.INTERNET_MNT_PATH is not None:
+        path = os.path.relpath(path, start=settings.INTERNET_MNT_PATH)
```
[the remark this answers is not in the document]
Comment (cytadela8) [reply wwwapp/views.py]: Is it? I'm actually not sure what will be supplied by nginx and was planning to adjust on production when we will know how this works exactly.

[the remark this answers is not in the document]
Comment (cytadela8) [reply wwwapp/views.py]: Also we can probably adjust nginx config so only the path is passed.

[the remark this answers is not in the document]
Comment (cytadela8) [reply wwwapp/views.py]: Not sure what will be easier.

File: wwwapp/views.py [hunk 6]
```diff
@@ -588,3 +595,31 @@ def page(request):
 
 index_view = as_article("index")
 template_for_workshop_page_view = as_article("template_for_workshop_page")
+
+
+@login_required()
+def resource_auth_view(request):
```
[the remark this answers is not in the document]
Comment (cytadela8) [reply wwwapp/views.py]: I don't know how to use this endpoint exactly :P Was planning to test on production. I have this: https://docs.nginx.com/nginx/admin-guide/security-controls/configuring-subrequest-authentication/

File: wwwapp/models.py [hunk 7]
```diff
@@ -22,6 +21,15 @@ class UserProfile(models.Model):
     def is_participating_in(self, year):
         return self.status_for(year) == 'Z' \
                or Workshop.objects.filter(type__year=year, lecturer=self, status='Z').exists()
+
+    def participation_years(self):
```
[the remark this answers is not in the document]
Comment (cytadela8) [reply wwwapp/models.py]: If we are fast enough we can merge this without type specification before grzegorz sees it :P

File: wwwapp/views.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (krzys-h) [reply wwwapp/views.py]: That's why I asked about the config - if you follow the nginx docs without any modification, this header will be set to `$request_uri`, which for a request to `http://warsztatywww.pl/internet/www15/file.txt?some=thing` will be `/internet/www15/file.txt?some=thing`

File: wwwapp/settings.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (krzys-h) [reply wwwapp/settings.py]: I bet you could configure it to not allow symlinks across resource directories. Or just disable symlinks altogether. Trying to implement such protections in a different part of the app than the one handling the request is usually just asking for trouble. What if there is some quirk where Python parses paths differently from nginx? Assuming your current code for parsing the request URI you can trivially bypass all checks with `/internet/www15/symlink?anything=xd` - python will look for a file called `symlink?anything=xd` and not find it while nginx will strip out the query string and serve `symlink`

[the remark this answers is not in the document]
Comment (krzys-h) [reply wwwapp/settings.py]: Oh, I think I get it now - you want to configure nginx to give you the full file path AFTER going though symlinks and then verify permissions. I guess that makes sense and should be reasonably secure, if configured properly.

[the remark this answers is not in the document]
Comment (cytadela8) [reply wwwapp/settings.py]: Yes, the idea was for nginx to give me full filepath so I can validate if this file can be accessed by currently loggedin user.

File: wwwapp/views.py [hunk 8]
```diff
@@ -588,3 +595,31 @@ def page(request):
 
 index_view = as_article("index")
 template_for_workshop_page_view = as_article("template_for_workshop_page")
+
+
+@login_required()
```
Comment (cytadela8) [inline wwwapp/views.py]: We might need to adjust it so permission denied is returned on logged out user. I'm more and more temted to test this code on production as I don't have a proper setup with nginx for development.

File: wwwapp/admin.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (grzegorz225) [reply wwwapp/admin.py]: We should refactor it in a separate PR.

File: wwwapp/admin.py [hunk 9]
```diff
@@ -51,3 +52,5 @@ def make_clear(self, _request, queryset):
 admin.site.register(WorkshopType)
 admin.site.register(WorkshopParticipant)
 admin.site.register(WorkshopUserProfile)
+
```
Comment (grzegorz225) [inline wwwapp/admin.py]: Why leave a vertical space here?

File: wwwapp/models.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (grzegorz225) [reply wwwapp/models.py]: Add documentation and type specification. Use mypy(merged to master) to check your types, BTW: Pycharm does some type checking for you.

File: wwwapp/models.py [hunk 10]
```diff
@@ -22,6 +21,15 @@ class UserProfile(models.Model):
...
+        return self.participant_years().union(self.lecturer_years())
+
+    def participant_years(self):
+        return set([profile.year for profile in self.workshop_profile.filter(status=WorkshopUserProfile.STATUS_ACCEPTED)])
+
+    def lecturer_years(self):
```
Comment (grzegorz225) [inline wwwapp/models.py]: Add documentation and types

File: wwwapp/models.py [hunk 11]
```diff
@@ -22,6 +21,15 @@ class UserProfile(models.Model):
     def is_participating_in(self, year):
         return self.status_for(year) == 'Z' \
                or Workshop.objects.filter(type__year=year, lecturer=self, status='Z').exists()
+
+    def participation_years(self):
+        return self.participant_years().union(self.lecturer_years())
+
+    def participant_years(self):
```
Comment (grzegorz225) [inline wwwapp/models.py]: Add documentation and types

File: wwwapp/models.py [hunk 12]
```diff
@@ -281,3 +289,23 @@ def result_in_percent(self):
 
     class Meta:
         unique_together = [('workshop', 'participant')]
+
+
+class ResourceYearPermission(models.Model):
```
Comment (grzegorz225) [inline wwwapp/models.py]: Documentation

Comment (grzegorz225) [reply wwwapp/models.py]: What about changing the name to CampResourcePermission - we will probably want to introduce a Camp model soon in order to avoid hardcoding the dates and the year in the settings.

File: wwwapp/models.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (grzegorz225) [reply wwwapp/models.py]: @krzys-h We should probably only use sizes which are powers of 2 :)

File: wwwapp/settings.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (grzegorz225) [reply wwwapp/settings.py]: Can we make sure that the resources are hosted on the same server as the app?

File: wwwapp/urls.py [hunk 13]
```diff
@@ -46,6 +46,7 @@
     url(r'^template_for_workshop_page/$', views.template_for_workshop_page_view, name='template_for_workshop_page'),
     url(r'^program/$', RedirectView.as_view(url='/%d/program/' % CURRENT_YEAR, permanent=False), name='program'),
     url(r'^([0-9]+)/program/$', views.program_view, name='year_program'),
+    url(r'^resource_auth/$', views.resource_auth_view),
```
Comment (grzegorz225) [inline wwwapp/urls.py]: Add name

File: wwwapp/views.py [hunk 14]
```diff
@@ -588,3 +595,31 @@ def page(request):
...
+def resource_auth_view(request):
+    if request.user.has_perm('wwwapp.access_all_resources'):
+        return HttpResponse("Glory to WWW and the ELITARNY MIMUW!!!")
```
[the remark this answers is not in the document]
Comment (grzegorz225) [reply wwwapp/views.py]: Did you mean ELYTARNY MIMUW?

File: wwwapp/views.py (hunk 8 above)
Comment (krzys-h) [reply wwwapp/views.py]: Don't worry, I'll definitely set ot up myself for testing before I give you an approve XD

Comment (krzys-h) [reply wwwapp/views.py]: Yeah, you definitely do
```
2019/07/22 10:43:08 [error] 11487#11487: *23 auth request unexpected status: 302 while sending to client, client: ::1, server: , request: "GET /internet/ HTTP/1.1", host: "localhost:8080"
::1 - - [22/Jul/2019:10:43:08 +0200] "GET /internet/ HTTP/1.1" 500 588 "-" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.100 Safari/537.36"
```

File: wwwapp/models.py (hunk 12 above)
Comment (krzys-h) [reply wwwapp/models.py]: I don't actually like either of these names - nothing prevents you from having multiple resources per year/camp and we could possibly introduce public resources at some point (e.g. photos). I would just call it `ResourcePermission` or even simply `Resource` (since it's not only about the permissions but also about displaying it in the menu)

File: wwwapp/admin.py (hunk 9 above)
Comment (krzys-h) [reply wwwapp/admin.py]: Because other things are in groups like this, probably :P

File: wwwapp/models.py [hunk 15]
```diff
@@ -281,3 +289,23 @@ def result_in_percent(self):
...
+    def __str__(self):
+        return "{} - {}".format(self.year, self.path)
+
+    def clean(self):
+        super().clean()
+        if self.access_url != "" and self.display_name == "":
+            raise ValidationError("Wyświetlana nazwa musi być ustawiona jeśli "
+                                  "URL dostępu jest ustawiony")
+
+    class Meta:
```
Comment (krzys-h) [inline wwwapp/models.py]: Specifying default ordering here may be a good idea
image

**Comment by grzegorz225** [comment]
We should create a docker environment  for testing this feature.

File: wwwapp/models.py [hunk 16]
```diff
@@ -22,6 +21,27 @@ class UserProfile(models.Model):
     def is_participating_in(self, year):
         return self.status_for(year) == 'Z' \
                or Workshop.objects.filter(type__year=year, lecturer=self, status='Z').exists()
+
+    def all_participation_years(self) -> list:
```
Comment (krzys-h) [inline wwwapp/models.py]: inb4 @grzegorz225 "not enough type info, should be List[int]" (or something like that, I don't remember :P)

File: wwwapp/views.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (krzys-h) [reply wwwapp/views.py]: Still missing proper docs on how to use this and what "path prefixes" are checked exactly

File: wwwapp/models.py [hunk 17]
```diff
@@ -281,3 +301,28 @@ def result_in_percent(self):
...
+
+class ResourceYearPermission(models.Model):
+    """
+    Resource asociated with a WWW edition (year). Resource can be accessed by
+    users who qualified or had a lecture on year.
+    """
+    display_name = models.CharField(max_length=50, blank=True)
+    access_url = models.URLField(blank=True,
+                                 help_text="URL dla przycisku w menu. Przycisk nie jest wyświetlany jeśli url jest pusty")
+    root_url = models.CharField(max_length=256, null=False, blank=False,
```
Comment (krzys-h) [inline wwwapp/models.py]: I don't like this name (this is not an URL, only a part of it), and isn't access_url basically root_url + hostname?

File: wwwapp/models.py [hunk 18]
```diff
@@ -281,3 +301,28 @@ def result_in_percent(self):
...
+                                help_text="URL bez protokołu. bez \"/\" na końcu.")
```
Comment (krzys-h) [inline wwwapp/models.py]: What about the beginning?

File: wwwapp/models.py (hunk 15 above)
Comment (cytadela8) [reply wwwapp/models.py]: Maybe year then name?

File: wwwapp/models.py (hunk 17 above)
Comment (cytadela8) [reply wwwapp/models.py]: root_url contains also hostname and port. (like warsztatywww.pl/asdasd) `access_url` is a proper URL for link. Maybe I could add an example to the comment.

File: wwwapp/urls.py (hunk 13 above)
Comment (cytadela8) [reply wwwapp/urls.py]: Name is for referencing in other parts of code. This does not need to be referenced.

File: wwwapp/models.py (hunk 12 above)
Comment (cytadela8) [reply wwwapp/models.py]: I would separate Resource and it's permissions in the future. Introducing public resources and complicated permissions is not really needed right now. This is really only designed for "INTERNETY". Even the menu dropdown title is `INTERNETY`. Public resources are and probably will in the future be hosted separately.

**Comment by cytadela8** [comment]
I created new commits for easier reviewing.

File: wwwapp/models.py (hunk 12 above)
Comment (zielmicha) [reply wwwapp/models.py]: Please don't use word "camp" as a name for the workshops :)

File: wwwapp/models.py (hunk 17 above)
Comment (krzys-h) [reply wwwapp/models.py]: Oh, I didn't notice you changed it (that's why this needed better docs :P). Why not a full URL then, it's confusing to have a hostname without http://

Are you even checking it anywhere? X-Original-URI doesn't have the hostname in default configuration iirc

In that case, maybe public_url and private_url or access_url and internal_url (the second one sounds better imho)?

File: wwwapp/models.py (hunk 15 above)
Comment (krzys-h) [reply wwwapp/models.py]: Sounds good

File: wwwapp/urls.py (hunk 13 above)
Comment (krzys-h) [reply wwwapp/urls.py]: But every single other endpoint has a name, including dataForPlan which doesn't need to be referenced anywhere either.

File: wwwapp/views.py [hunk 19]
```diff
@@ -588,3 +593,46 @@ def page(request):
...
+
+
+def resource_auth_view(request):
+    """
+    View checking permission for resource (header X-Original-URI). Returns 200
+    when currently logged in user should be granted access to resource and 403
+    when access should be denied.
+
+    See https://docs.nginx.com/nginx/admin-guide/security-controls/configuring-subrequest-authentication/
+    for intended usage.
```
Comment (krzys-h) [inline wwwapp/views.py]: I'm still missing info on how exactly are the paths processed (which paths will be available when you give access to e.g. `/internety/www15`, and which paths are checked when you try to acess some URL [all prefixes])

File: wwwapp/models.py [hunk 20]
```diff
@@ -281,3 +302,30 @@ def result_in_percent(self):
...
+    """
+    Resource associated with a WWW edition (year). Resource can be accessed by
+    users who qualified or had a lecture on year. User is granted access to
+    root_url and recursively to all files and subdirectories inside.
+    """
+    display_name = models.CharField(max_length=50, blank=True)
+    access_url = models.URLField(blank=True,
+                                 help_text="URL dla przycisku w menu. Przycisk nie jest wyświetlany jeśli url jest pusty")
+    root_url = models.CharField(max_length=256, null=False, blank=False,
+                                help_text='URL bez protokołu, bez "/" na końcu. np. "internet.warsztatywww.pl/www15"')
```
Comment (krzys-h) [inline wwwapp/models.py]: I would add validation of the format in clean(), especially the trailing / rule because that's easy to miss

File: wwwapp/views.py (hunk 19 above)
Comment (cytadela8) [reply wwwapp/views.py]: Such info is in the model docs.

File: wwwapp/models.py (hunk 17 above)
Comment (cytadela8) [reply wwwapp/models.py]: So what does X-Original-URI actually contain? At first I implemented for a case where it's only a path, now i implemented for URL.

File: wwwapp/views.py (hunk 19 above)
Comment (cytadela8) [reply wwwapp/views.py]: Such info is in the model.

File: wwwapp/models.py (hunk 17 above)
Comment (krzys-h) [reply wwwapp/models.py]: Well... like I said a few times already, it's the $request_uri variable from nginx which is the request URI part of the request as defined in [RFC 2616](https://www.w3.org/Protocols/rfc2616/rfc2616-sec5.html#sec5.1.2) :P

Tl;dr it's `/internet/www14/something?query=string` (the path and query string parts of the URL)

Comment (cytadela8) [reply wwwapp/models.py]: So it was implemented properly before...

Comment (krzys-h) [reply wwwapp/models.py]: It wasn't - you were trying to interpret this URI a a file path in the filesystem with INTERNET_MNT_PATH. The code path where this variable was None was correct though.

Comment (krzys-h) [reply wwwapp/models.py]: Or actually, it wasn't either, it would break with query strings too. The current implementation is better if you just get rid of the hostname. It may be a good idea to make root_url always start with a `/` to make it look more like an actual request URI, although that's not required from functionality point of view.

File: wwwapp/models.py [hunk 21]
```diff
@@ -281,3 +305,49 @@ def result_in_percent(self):
...
+    """
+    Resource associated with a WWW edition (year). Resource can be accessed by
+    users who qualified or had a lecture on year. User is granted access to
+    root_url and recursively to all files and subdirectories inside.
+    """
+    display_name = models.CharField(max_length=50, blank=True)
+    access_url = models.URLField(blank=True,
+                                 help_text="URL dla przycisku w menu. Przycisk nie jest wyświetlany jeśli url jest pusty")
+    root_uri = models.CharField(max_length=256, null=False, blank=False,
+                                help_text='bez "/" na końcu i na początku. np. "internety/www15"')
```
Comment (krzys-h) [inline wwwapp/models.py]: I would prefer this to always start with a `/` so that it looks like a proper URI

Comment (cytadela8) [reply wwwapp/models.py]: I will rename to "path". I checked and URI seems to be a set of strings containg all url-s.

Comment (krzys-h) [reply wwwapp/models.py]: The path part of the URI should still start with a `/` :P

```
For the URI http://www.example.org:56789/a/b/c.txt?t=win&s=chess#para5 these would be:

Scheme: http
Authority: www.example.org:56789
  User Information: not present
  Host: www.example.org
  Port: 56789
Path: /a/b/c.txt
Query: t=win&s=chess
Fragment: para5
```

File: wwwapp/models.py [hunk 22]
```diff
@@ -281,3 +305,49 @@ def result_in_percent(self):
...
+
+    @staticmethod
+    def resources_for_uri(uri: str):
+        scheme, netloc, path, query, fragment = urllib.parse.urlsplit(uri)
+        path = os.path.normpath(path)  # normalize path
+        path_parts = path.split('/')
+
+        query = Q(pk__isnull=True)  # always false
+        for i in range(len(path_parts)):
+            query |= Q(root_path='/'.join(path_parts[:i+1]))
```
Comment (krzys-h) [inline wwwapp/models.py]: This code checks the following prefixes:
```
(empty string)
/internet
/internet/www13
```
The first one should be `/`, not empty string.

Comment (cytadela8) [reply wwwapp/models.py]: Ok! More IFs INCOMING!!!!

**Comment by cytadela8** [comment]
Rebased

File: wwwapp/migrations/0057_resourceyearpermission.py [hunk 23]
```diff
@@ -0,0 +1,29 @@
...
+    ]
+
+    operations = [
+        migrations.CreateModel(
+            name='ResourceYearPermission',
+            fields=[
+                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
+                ('display_name', models.CharField(blank=True, max_length=50)),
+                ('access_url', models.URLField(blank=True, help_text='URL dla przycisku w menu. Przycisk nie jest wyświetlany jeśli url jest pusty')),
+                ('root_path', models.CharField(help_text='bez "/" na końcu i na początku. np. "internety/www15"', max_length=256)),
```
Comment (krzys-h) [inline wwwapp/migrations/0057_resourceyearpermission.py]: Update the migration for new help_text (I don't think it really matters though)

File: wwwapp/models.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (krzys-h) [reply wwwapp/models.py]: Now you could do it, the other constant is added ;)

**Comment by cytadela8** [comment]
There is more refactoring required for the Workshop state constant.

File: wwwapp/models.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (cytadela8) [reply wwwapp/models.py]: There are other cases of constants not being used. I will fix this in a separate refactoring PR.

File: nginx.conf.example [hunk 24]
```diff
@@ -0,0 +1,85 @@
+# nginx -p ./ -c nginx.conf
+
+# display_name=Internet WWW11, access_url=http://localhost:8080/internet/www11, path=internet/www11, year=2015
+# display_name=Internet WWW12, access_url=http://localhost:8080/internet/www12, path=internet/www12, year=2016 - nie ma plików bo zaginęły :P
```
Comment (grzegorz225) [inline nginx.conf.example]: :(

File: nginx.conf.example [hunk 25]
```diff
@@ -0,0 +1,85 @@
...
+# display_name=Internet WWW14, access_url=http://localhost:8080/internet/www14, path=internet/www14, year=2018
+
+worker_processes auto;
+daemon off;
+pid nginx.pid;
+error_log /dev/stdout;
+
+events {
+	worker_connections 768;
+	# multi_accept on;
```
Comment (grzegorz225) [inline nginx.conf.example]: ?

File: wwwapp/models.py [hunk 26]
```diff
@@ -301,3 +323,52 @@ def result_in_percent(self):
 
     class Meta:
         unique_together = [('workshop', 'participant')]
+
+
+class ResourceYearPermission(models.Model):
+    """
+    Resource associated with a WWW edition (year). Resource can be accessed by
+    users who qualified or had a lecture on year. User is granted access to
```
Comment (grzegorz225) [inline wwwapp/models.py]:
```suggestion
    users who qualified or had a lecture on a given year. User is granted access to
```

File: wwwapp/views.py [hunk 27]
```diff
@@ -631,3 +634,27 @@ def page(request):
 
 index_view = as_article("index")
 template_for_workshop_page_view = as_article("template_for_workshop_page")
+
+
+def resource_auth_view(request):
+    """
+    View checking permission for resource (header X-Original-URI). Returns 200
```
Comment (grzegorz225) [inline wwwapp/views.py]:
```suggestion
    View checking permission for resource (header X-Original-URI). Returns an HTML response with status code 200
```

File: wwwapp/views.py [hunk 28]
```diff
@@ -631,3 +634,27 @@ def page(request):
...
+    when currently logged in user should be granted access to resource and 403
```
Comment (grzegorz225) [inline wwwapp/views.py]:
```suggestion
    when currently logged in user should be granted access to resource and a response with status code 403
```

File: wwwapp/views.py [hunk 29]
```diff
@@ -631,3 +634,27 @@ def page(request):
...
+    when access should be denied.
```
Comment (grzegorz225) [inline wwwapp/views.py]:
```suggestion
    when access to the resource should be denied.
```

Comment (cytadela8) [reply wwwapp/views.py]: Nah

File: nginx.conf.example (hunk 25 above)
Comment (cytadela8) [reply nginx.conf.example]: @krzys-h file. This example is not really important. Probably this is a default (so uncommenting wouldn't change anything.

Comment (krzys-h) [reply nginx.conf.example]: This is just a heavily modified default config from Ubuntu, I left everything that wasn't important as-is
